### PR TITLE
Added user status and it's timestamp storage for GTM tracking.

### PR DESCRIPTION
**Pull request by juliquiron**
Stores data to use for GTM tracking

### CODE REVIEW

**Review by rodrigoaguilera — APPROVED** [review]

File: ssofact.module [hunk 1]
```diff
@@ -35,23 +35,216 @@ function ssofact_help($route_name, RouteMatchInterface $route_match) {
...
+  // unset($properties_ignore[array_search('changed', $properties_ignore)]);
 }
 
 /**
  * Implements hook_openid_connect_userinfo_save().
  */
 function ssofact_openid_connect_userinfo_save(UserInterface $account, $context) {
-  $userinfo = $context['userinfo'];
-  if (empty($userinfo['deactivated'])) {
+  $user_info = $context['userinfo'];
```
Comment (sun) [inline ssofact.module:45]: Why did you change the name here?  The variable is named `$userinfo` in the rest of the code.  It's also not consistent with the key in `$context`.  Renaming it here is bad for discovery/code-search.

File: ssofact.module [hunk 2]
```diff
@@ -35,23 +35,216 @@ function ssofact_help($route_name, RouteMatchInterface $route_match) {
...
+        $active_purchases[] = $access['purchase'];
+      }
+      else {
+        $expired_purchases[] = $access['purchase'];
+      }
+    }
+
+    // No active subscriptions but article test active.
+    if (empty($active_purchases) && !$article_test->hasUserAccessToNextTest($account)) {
+      $customer_status = 'trial | article';
```
Comment (sun) [inline ssofact.module:92]: Didn't we discuss to store the _"columns"_ of the defined tracking values as separate user data rows?

I did not expect to see pipes (`|`) in the user data values.  Those are the column separators.  The values of the columns should come from different user data rows.

The expected user data rows / values use only one column value, as we discussed and documented in the Asana ticket:
```
customer_status: trial
trial_access: article
subscription_product: ''
last_purchase_timestamp: 123456789
```
The tracking module code then concatenates those values to get the desired tracking value, along the lines of the following:
```php
$status = implode(' | ', array_filter([$customer_status, $trial_access, $subscription_product]));
$age = floor(($time - $last_purchase_timestamp) / 86400);
```

Storing the final tracking values instead of the atomic meanings makes the stored values not reusable for anything else (because we cannot hardcode the special meanings of tracking parameters with weird pipes in them into other places).

File: ssofact.module [hunk 3]
```diff
@@ -35,23 +35,216 @@ function ssofact_help($route_name, RouteMatchInterface $route_match) {
...
+        $last_purchase_timestamp = $free_trial_timestamp;
+      }
+      elseif ($have_free_trials) {
+        $customer_status = 'ex | trial | paid';
+        $last_purchase_timestamp = $paid_trial_timestamp;
+      }
+    }
+    // There are active purchases.
+    else {
+      extract_purchases_info($active_purchases, $have_free_trials, $have_paid_trials, $have_subscriptions, $product_labels, 'fromDay', $free_trial_timestamp, $paid_trial_timestamp, $subscription_timestamp);
```
Comment (sun) [inline ssofact.module:118]: 1. Function name is not in module namespace.

1. Minor: A long list of parameters passed by reference is a quite weird coding pattern (if it can be called a pattern at all).  Consider to use a single associative array instead.  We can change this later as long as this is internal and no other code calls the function.

**Review by sun — CHANGES_REQUESTED** [review]
I'm confused 😕 This is not really the implementation we had discussed?  If it helps we can have a call about this.

**Comment by juliquiron** [comment]
As discussed in person, this code should be moved to `nepo_custom` for the meanwhile. At some point refactor to a unified management of the customer journey status, followup task: https://app.asana.com/0/633061025239778/815505249992665

**Comment by juliquiron** [comment]
Moved code to: https://github.com/netzstrategen/nepo/pull/158/commits/aefd5a707526b80d273b4751b3a9ab19c7959306
introduced some comments as todo for further refactoring.

THIS PR SHOULD NOT BE MERGED. closeing.

**Comment by sun** [comment]
👍 Thanks!
